File: 572. Subtree of Another Tree.py
# 572. Subtree of Another Tree


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not subRoot:
            return True
        if not root:
            return False

        # at this point we;re sure both tree exists in the very beginning
        def find(l, r):
            if not l and not r:
                return True
            if not l or not r:
                return False
            if l.val != r.val:
                return False

            return find(l.left, r.left) and find(l.right, r.right)

        return (
            find(root, subRoot)
            or self.isSubtree(root.left, subRoot)
            or self.isSubtree(root.right, subRoot)
        )

File: 572. Subtree of Another Tree.cpp
// 572. Subtree of Another Tree


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool same(TreeNode* l, TreeNode* r) {
        if (!l && !r) { return true; }
        if (!l || !r) { return false; }
        if (l->val != r->val) { return false; }
        return same(l->left, r->left) && same(l->right, r->right);
    }
    bool res {false};
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!subRoot) {return true;}
        if (!root) {return false;}
        if (root->val != subRoot->val) {
            res = (isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot));
        }
        else {
            return res = (isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot) ||same(root, subRoot));
        }
        
        return res;

        
    }
};